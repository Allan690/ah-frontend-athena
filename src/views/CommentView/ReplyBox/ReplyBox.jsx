import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { CommentAction } from "../../../actions/CommentActions/CommentAction";
import { CommentEditAction } from "../../../actions/CommentActions/CommentEditAction";
import {
  ReplyPostAction,
  CommentGetAction
} from "../../../actions/CommentActions/CommentGetAction";
import CommentCard from "../../../components/Comments/CommentCard/CommentCard";

let token = window.localStorage.getItem("token");

export class ReplyBox extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  onSubmitReply = id => {
    const commentobj = this.state;

    if (!token) {
      return null;
    } else {
      const comment = {
        comment_body: commentobj.comment_body
      };
      const {
        ReplyPostAction,
        CommentEditAction,
        CommentAction,
        art_slug,
        isEdit
      } = this.props;
      if (isEdit) {
        if (id === null) {
          CommentAction(comment, art_slug);
        } else {
          CommentEditAction(id, art_slug, comment);
        }
      } else {
        const { parentId, art_slug } = this.props;
        if (parentId && art_slug) {
          ReplyPostAction(id, comment, art_slug);
        } else {
          CommentAction(comment, art_slug);
          console.log(comment, art_slug)
          
        }
      }
    }
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };
  onSubmit = e => {
    e.preventDefault();
    const { childId } = this.props;
    this.onSubmitReply(childId);
  };
  render() {
    const { childId} = this.props;
    return (
      <CommentCard
        childId={childId}
        onChange={this.onChange}
        onSubmit={this.onSubmit}
        onSubmitReply={this.onSubmit}
      />
    );
  }
}

ReplyBox.propTypes = {
  CommentAction: PropTypes.func,
  ReplyPostAction: PropTypes.func,
  CommentEditAction: PropTypes.func,
  art_slug: PropTypes.string,
  isEdit: PropTypes.bool,
  parentId: PropTypes.number,
  childId: PropTypes.number
};
ReplyBox.defaultProps = {
  CommentAction: () => {},
  ReplyPostAction: () => {},
  CommentEditAction: () => {},
  art_slug: "",
  isEdit: false,
  parentId: null,
  childId: null
};
export const mapStateToProps = state => {
  return {
    currentUser: state.login.login,
    art_slug: state.articles.view_article.slug
  };
};

export default connect(
  mapStateToProps,
  { CommentAction, ReplyPostAction, CommentGetAction, CommentEditAction }
)(ReplyBox);
