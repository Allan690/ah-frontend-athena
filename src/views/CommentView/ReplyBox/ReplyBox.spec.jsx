import React from "react";
import { shallow, mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { ReplyBox, mapStateToProps } from "./ReplyBox";
import { CommentAction } from "../../../actions/CommentActions/CommentAction";
import { CommentEditAction } from "../../../actions/CommentActions/CommentEditAction";
import {
  ReplyPostAction,
  CommentGetAction
} from "../../../actions/CommentActions/CommentGetAction";

let props = {
  mockSignUpFn: jest.fn(),
  isEdit: false, 
};
const initialState = {
  comment: {
    commentList: {},
    comment: {},
    replyList: [],
    refresh: false
  },
  login: {
    login: {
      username: "kasule",
      email: "okello@andela.com",
      password: "",
      errors: ""
    }
  },
  isEdit: false,
  articles: {
    view_article: {
      slug: "someslug"
    }
  }
};
let store;
const mockStore = configureStore();
store = mockStore(initialState);
let wrapper;
const defaultProps ={
  isEdit: false,
  CommentAction: ()=>{},
  CommentEditAction: ()=>{},
  ReplyPostAction:()=>{},
  onSubmitReply: ()=>{},
  onSubmit: ()=>{},
  onChange: ()=>{},
}
const updateProps = props => Object.assign(defaultProps, props);
describe("test ReplyBox view", () => {
  
  
  beforeEach(() => {
    wrapper = shallow(<ReplyBox {...defaultProps} />);
  });
  it("should  ReplyBox match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("test handle onSubmitReply event", () => {
    jest.setTimeout(10000);
    wrapper.setProps({
      isEdit: true
    });
    wrapper.instance().onSubmitReply();

    expect(wrapper.state()).toBe(true);
  });

  it("should handle onchange function", () => {
    wrapper
      .instance()
      .onChange({ target: { name: "comment", value: "my comment" } });
    expect(wrapper.state("comment")).toEqual("my comment");
  });
  it("should handle onsubmit function", () => {
    wrapper.instance().onSubmit({ preventDefault() {} });
    expect(wrapper.state()).toEqual({});
  });
  it("should handle onSubmitReply function", (id = jest.fn()) => {
    wrapper.instance().onSubmitReply(id);
    wrapper.setState({
      isEdit: true
    });
    expect(wrapper.state("isEdit")).toEqual(true);
  });

  it("should recieve props", () => {
    const wrapper2 = mount(
      <Provider store={store}>
        <ReplyBox {...props} />
      </Provider>
    );
    wrapper2.setState({
      isEdit: true
    });
    console.log(wrapper2.instance());

    expect(wrapper2.instance()).toBeCalled(CommentAction);
    expect(wrapper2.state()).toEqual({
      errors: "username should be between 3 to 8 characters"
    });
  });
  it("should map state to props", () => {
    expect(mapStateToProps(initialState).currentUser["username"]).toEqual(
      "kasule"
    );
  });
  it("should test prototype of data", () => {
    const data = {}
    const slug = 'hdj'
    const wrapper2 = mount(
      <Provider store={store}>
        <ReplyBox {...updateProps}  />
      </Provider>
    );
    expect(wrapper2.instance()).toEqual('');
  });
});
