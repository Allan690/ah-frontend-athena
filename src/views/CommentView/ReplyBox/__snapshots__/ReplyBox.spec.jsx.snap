// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ReplyBox view should  ReplyBox match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReplyBox
    CommentAction={[Function]}
    CommentEditAction={[Function]}
    ReplyPostAction={[Function]}
    art_slug=""
    childId={null}
    isEdit={true}
    parentId={null}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "childId": null,
      "onChange": [Function],
      "onSubmit": [Function],
      "onSubmitReply": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "childId": null,
        "onChange": [Function],
        "onSubmit": [Function],
        "onSubmitReply": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`test ReplyBox view should test prototype of data 1`] = `
<ReplyBox
  CommentAction={[Function]}
  CommentEditAction={[Function]}
  ReplyPostAction={[Function]}
  childId={null}
  isEdit={false}
  mockSignUpFn={[MockFunction]}
  parentId={null}
  slug=""
/>
`;
